flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -403,7 +403,10 @@ public void shutdown(JobStatus jobStatus) throws Exception {
 			// clear queued requests and in-flight checkpoints
 			abortPendingAndQueuedCheckpoints(reason);
 
-			completedCheckpointStore.shutdown(jobStatus);
+			// there might be a race condition with IO threads on completedCheckpointStore
```

Comment:
The race condition is between io threads and just the shut down? And previously was it working becase `CompletedCheckpointStore` access were synchronized on the checkopint coordinator's lock?

Reply:
The race condition happens between `completedCheckpointStore.shutdown` and `completedCheckpointStore.addCheckpoint`.

> And previously was it working becase CompletedCheckpointStore access were synchronized on the checkopint coordinator's lock?

Yes, exactly.